### DISCUSSION TITLE
v0.3.3 - AbstractIntegratedModule binary New Release

### DISCUSSION THREAD

**Discussion by Micro-Novelty** [Micro-Novelty/IntegratedPipeline-Specialized-Non-LLM-AI-Agent-Framework#94] [General]
### Note: the precompiled Wheels for aarch64 and Windows architecture with additional tarball, is now officially provided in the library! meaning the binaries provided here will act as backup for those who can't download the pip version.
 - Or consider using this specific pip Installation method for getting your specific Wheel tied to your Python version and OS:
- ```bash
    pip install abstractintegratedmodule --extra-index-url https://Micro-Novelty.github.io/abstract-modules/whl/ --break-system-packages
   # ensures proper installation by bypassing pip strict external download setup. 
   ```
   - Note: by using this Method, it will automatically Download the specific Wheel that matched Your python version and Your OS, This installation method will provide you with the Latest version of AbstractIntegratedModule.

## 🎯 Overview
IntegratedPipeline is a standalone Custom AI Agent Library for memory-augmented agentic frameworks, designed for continuous learning with MANN-Type Architecture designed for long term operation.

## 🎯 The provided Files below includes:
- Binaries for all AbstractIntegratedmodule from .pyd to .so files. (In a zipped package file)
- The provided binaries below supports python 3.10 - 3.12 for each Windows, Mac-Os and Linux x86-64 architecture.
- Contain binaries for:
   - Windows x86-64 architecture
   - Mac-Os architecture.
   - Linux x86-64 architecture.

## ✨ What's New and Included
- Added Rust specific compiled binary for handling complex weight matrices.
          - added specific optimized weight saving handling with rust optimization.
          - Adding Optimization and refinements, especially bug fixes in:
             - IntegratedPipeline lstm samples creating
             - IntegratedPipeline shape adaptation
             - IntegratedPipeline probability calibration
             - IntegratedPipeline auto generate labels text
             - IntegratedPipeline MLP samples generation.
             
### Core Features
- **Memory-Augmented Neural Networks (MANN)** - External dynamic memory module for persistent learning
- **Abstract Weight Encoder (AWE)** - Eigenvalue-based weight shaping for noise robustness
- **Specialized MLP** - Noise-robust classification with geometric weight adaptation
- **Optimized Transformer** - Alpha-based computing for stable contextual understanding
- **LSTM Architectures** - Efficient LSTM Using AWE Weight in LSTM  __init__.
- **Hybrid Ensemble** - Dynamic weighting of MLP + Transformer predictions
- **SQLite Integration** - Local database for memory, attention weights, and predictions
- **Peer-to-Peer Coordination** - Distributed agent communication with SSL security, Cohesive Interaction, and Compact multi-modal ensemble weighting from peer.

### Platform Support
- ✅ Windows (x86_64)
- ✅ Linux (x86_64)
- ✅ Raspberry Pi (ARM64 v7/v8) 

### Python Support
- Python 3.13 Only for windows 
- Python 3.12 for x86-64 machine
- Python 3.10 for ARM64 machine (Stable Robust version, Officially tested in ARM64 Environment)

##  📦 Library requirements when using binaries
- numpy
- scikit-learn
- pandas
- aiohttp
- psutil

## 📦 Binaries Included

Download the appropriate binary for your platform:
- `AbstractIntegratedModule.pyd` and 'AbstractOptimizedModules.pyd' - Windows
- `AbstractIntegratedModule.cpython-39-x86_64-linux-gnu.so` and 'AbstractOptimizedModules.cpython-312-x86_64-linux-gnu.so' - Linux x86_64
- `AbstractIntegratedModule.cpython-39-aarch64-linux-gnu.so` and 'AbstractIntegratedModule.cpython-310-aarch64-linux-gnu.so' - Raspberry Pi ARM64
- Note: All of this binaries are provided in the packaged zip.

Download the appropriate files too:
- `Dockerfile + required files (see in Docker Application block in this repo)` (For Container assemble), can be downloaded in the code section 
- `main.py` (Script to execute in Quick_Docker_start)
- `Quick_Docker_start.sh` (For quick start)


### Bug fixed:
- fixed undefined variables bug in calibrate probs 
- fixed nan fragility in trust scoring in P2P.
- fixed loading model dict that only returns None

## 🚀 Quick Start

```python
from AbstractIntegratedModule import IntegratedPipeline
from AbstractIntegratedModule import PipelinePredictionManager
import numpy as np

memory_name = 'agent_memory'
cert_file = <your_cert_file_dir> # your .crt file
key_file = <your_key_file_dir> # your .key file

main_model = IntegratedPipeline(memory_name=memory_name, use_async=True, agent_port=5000, ssl_cert_file=cert_file, ssl_key_file=key_file)
 # provide cert_file path or key_file path (optional)
main_prediction = PipelinePredictionManager(main_model, 
               label_csv='C:/users/yourdevice/example_manual_training.txt', 
                  # or /home/yourdevice/example_manual_training.txt.
                  #your path dir that contains the .txt file that contains CSV format.
               target_title='window_title', label='label')
# example_manual_training is a .txt file that contain csv format like above example.

example_rules = [
                     # === WORK / PRODUCTIVITY ===
                     (r'code|programming|develop|debug|compile|script', 'focused_work'),
                     (r'vscode|visual_studio|ide|terminal|shell', 'focused_work'),
                     (r'notion|evernote|onenote|notes|todo|task', 'productive'),
                     (r'slack|teams|discord|zoom|meeting|call', 'communication'),
                     (r'email|gmail|outlook|inbox|mail', 'communication'),
                     
                     # === ENTERTAINMENT ===
                     (r'youtube|netflix|twitch|stream|video', 'entertainment'),
                     (r'music|spotify|soundcloud|audio|player', 'entertainment'),
                     (r'game|gaming|steam|epic|play', 'gaming'),
                     (r'facebook|instagram|tiktok|social|post', 'social_media'),
                     
                     # === BROWSING ===
                     (r'chrome|firefox|edge|safari|browser', 'browsing'),
                     (r'google|search|wiki|wiki|article', 'information'),
                     (r'stackoverflow|github|docs|documentation', 'research'),

                     # more rules
                 ]
# activate explainability capability to explain uncertainty:
main_model.show_explainability_details = True
main_model.distribution.predict_manager = main_prediction # set PipelinePredictionManager to AgentDistributedInference for asynchronous prediction later

# test samples with more sophisticated rules and more complex titles for prediction
# (title, intent)
test_titles = [
 ("Opening Thesis.docx", "slight_work"),
 ("Watching YouTube and Google Chrome", "distracted"),
 ("Watching Slack", "communication"),
 ("Programming in Visual Studio Code", "focused_work"),
 ("Watching netflix.com - Chrome", "break"),
# more titles 
 ]  
            
titles, y, label_map = main_prediction.load_labels_from_csv(
    <C:/your/path/to/example_manual_training.txt>, # your actual .txt file that contains CSV format for trainings.  same like the abpv initialized label_csv directory, can also use the provided ManualsTraining.txt below for quick use.
    <target_title>, <target_label>)
# small training with simple titles
main_model.train(titles, y)
   
results, chosen_label, confidence = main_prediction.advanced_prediction_method(test_titles, label_map, example_rules,
                             X=None, y=None,
                             show_proba=False, top_k=3, 
                             use_transformer=True,
                             return_attention=False,
                             save_results=True)

# ... more features/wrapper you can add
```
)

<hr /><em>This discussion was created from the release <a href='https://github.com/Micro-Novelty/IntegratedPipeline-Specialized-Non-LLM-AI-Agent-Framework/releases/tag/v0.3.3-0.9.8'>v0.3.3 - AbstractIntegratedModule binary New Release</a>.</em>
